src/drivers/amd/common/acp_dma.c

```diff
@@ -0,0 +1,394 @@
+			/* clear the reset flag after successfully stopping the dma transfer
+			 * and break from the loop
+			 */
+			dma_cntl.bits.dmachrst = 0;
+			dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+			break;
+		}
+		delay_cnt--;
+	}
+	return 0;
+}
+
+static int acp_dma_start(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t dma_cntl;
+	acp_dma_ch_sts_t dma_sts;
+	uint32_t chan_sts;
+	int ret = 0;
+	struct timer *timer = timer_get();
+	uint64_t deadline = platform_timer_get(timer) +
+				clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) * 500 / 1000;
+	if (channel->status != COMP_STATE_PREPARE &&
+	    channel->status != COMP_STATE_SUSPEND)
+		return -EINVAL;
+	channel->status = COMP_STATE_ACTIVE;
+	/* Clear DMAChRun before starting the DMA Ch */
+	dma_cntl =  (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun = 0;
+	dma_cntl.bits.dmachiocen  = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	dma_cntl =  (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun	 = 1;
+	dma_cntl.bits.dmachiocen = 0;
+
+	/* set dmachrun bit to start the transfer */
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	/* poll for the status bit
+	 * to finish the dma transfer
+	 * then initiate call back function
+	 */
+	dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	chan_sts = dma_sts.u32all & (1 << channel->index);
+	while (chan_sts) {
+		if (deadline < platform_timer_get(timer)) {
+			/* safe check in case we've got preempted after read */
+			if (chan_sts)
+				return 0;
+			tr_err(&acpdma_tr, "acp-dma: timed out for dma start");
+			return -ETIME;
+		}
+		dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		chan_sts = dma_sts.u32all & (1 << channel->index);
+	}
+	return ret;
+}
+
+static int acp_dma_release(struct dma_chan_data *channel)
+{
+	tr_info(&acpdma_tr, "DMA: release(%d)", channel->index);
+	if (channel->status != COMP_STATE_PAUSED)
+		return -EINVAL;
+	channel->status = COMP_STATE_ACTIVE;
+	return 0;
+}
+
+static int acp_dma_pause(struct dma_chan_data *channel)
+{
+	tr_info(&acpdma_tr, "h/w pause is not supported, changing the status of(%d) channel",
+		channel->index);
+	if (channel->status != COMP_STATE_ACTIVE)
+		return -EINVAL;
+	channel->status = COMP_STATE_PAUSED;
+	return 0;
+}
+
+static int acp_dma_copy(struct dma_chan_data *channel, int bytes, uint32_t flags)
+{
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+	acp_dma_ch_sts_t ch_sts;
+	uint32_t dmach_mask = (1 << channel->index);
+	int ret = 0;
+
+	if (flags & DMA_COPY_ONE_SHOT) {
+		ret = acp_dma_start(channel);
+		if (ret < 0)
+			return ret;
+		ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		while (ch_sts.bits.dmachrunsts & dmach_mask) {
+			ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+			if (!(ch_sts.bits.dmachrunsts & dmach_mask))
+				break;
```

Comment:
> the `if (!x) break;` can be removed:
> 
> ```
> while (x) {
>   x = fn();
>   if (!x)
>     break;
> }
> ```

It is a valid condition. It will break when the DMA run status is false.

Reply:
sorry, you already check it in the loop. The loop will stop when the condition is false. You don't need to check it again. If you had an infinite loop you'd need a break, like
```
while (1) {
  x = fn();
  if (!x)
    break;
}
```
but your loop already uses the condition, so
```
while (x) {
  x = fn();
}
```
is perfectly enough

Reply:
After entering the while condition, the first statement updates the variable `ch_sts`. So checking again (In statement `if (!(ch_sts.bits.dmachrunsts & dmach_mask))` )  for breaking the loop.

Reply:
but you don't need that. Without that `if` the loop will jump back to line 222 `while (ch_sts.bits.dmachrunsts & dmach_mask) {` will check **exactly the same** condition as in your `if` and will terminate the loop too. That's exactly what I meant when I gave you this example:
```
while (x) {
  x = fn();
  if (!x)
    break;
}
```

Reply:
and spinning on a register is not a nice thing, doing that without timeout can lead to 
```c
timeout = 10;
while (timeout) {
        if(!func())
                break;
        sleep(1);
        timeout--;
}

if (!timeout)
        return -ETIMEDOUT;
```
there might be a macro for this around?

Reply:
Done the changes accordingly.

Regarding the timeout, We are in the progress of implementing the timeout functionality using the Watchdog timer. So that will be handled further.